packages/core/src/priority-queue/persisted.ts

```diff
@@ -0,0 +1,136 @@
+import { PriorityQueue } from '.'
+import { CoreContext, SerializedContext } from '../context'
+
+const nullStorage = (): Storage => ({
+  getItem: () => null,
+  setItem: () => null,
+  removeItem: () => null,
+  length: 0,
+  clear: () => null,
+  key: () => null,
+})
+
+function persisted(loc: Storage, key: string): CoreContext[] {
+  const items = loc.getItem(key)
+  return (items ? JSON.parse(items) : []).map(
+    (p: SerializedContext) => new CoreContext(p.event, p.id)
+  )
+}
+
+function persistItems(loc: Storage, key: string, items: CoreContext[]): void {
+  const existing = persisted(loc, key)
+  const all = [...items, ...existing]
+
+  const merged = all.reduce((acc, item) => {
+    return {
+      ...acc,
+      [item.id]: item,
+    }
+  }, {} as Record<string, CoreContext>)
+
+  loc.setItem(key, JSON.stringify(Object.values(merged)))
+}
+
+function seen(loc: Storage, key: string): Record<string, number> {
+  const stored = loc.getItem(key)
+  return stored ? JSON.parse(stored) : {}
+}
+
+function persistSeen(
+  loc: Storage,
+  key: string,
+  memory: Record<string, number>
+): void {
+  const stored = seen(loc, key)
+
+  loc.setItem(
+    key,
+    JSON.stringify({
+      ...stored,
+      ...memory,
+    })
+  )
+}
+
+function remove(loc: Storage, key: string): void {
```

Comment:
Not necessary to be in core (the tests weren’t as brittle as I thought), but I refactored this to no longer create any import side effects.